### PR TITLE
[FLINK-3780] [gelly] Jaccard Similarity

### PR DESCRIPTION
**Pull request by greghogan**
The Jaccard Index measures the similarity between vertex neighborhoods. Scores range from 0.0 (no common neighbors) to 1.0 (all neighbors are common).

### CODE REVIEW

**Comment by greghogan** [comment]
Tests are passing ... will merge ...

**Comment by vasia** [comment]
Hey @greghogan,
it's usually a good idea to have at least one person look at the code :)
Is this blocking some other feature? If not, I could take a look at this PR tomorrow if you like.

**Comment by greghogan** [comment]
Hi @vasia that would be great if you have time after 2044.

**Comment by vasia** [comment]
Great, that's my plan :)

**Comment by greghogan** [comment]
I figured. I forgot to mention that I was looking to reference the JaccardIndex self-join in the discussion of FLINK-3910 but I can link into the PR this time.

File: docs/apis/batch/libs/gelly.md [hunk 1]
```diff
@@ -2051,6 +2052,26 @@ The algorithm takes a directed, vertex (and possibly edge) attributed graph as i
...
 vertex and edge in the output graph stores the common group value and the number of represented elements.
 
+### Jaccard Index
+
+#### Overview
+The Jaccard Index measures the similarity between vertex neighborhoods. Scores range from 0.0 (no common neighbors) to
+1.0 (all neighbors are common).
+
+#### Details
+Counting common neighbors for pairs of vertices is equivalent to counting the two-paths consisting of two edges
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: By "two-paths" you mean triads? i.e. open triangles?

File: docs/apis/batch/libs/gelly.md [hunk 2]
```diff
@@ -2051,6 +2052,26 @@ The algorithm takes a directed, vertex (and possibly edge) attributed graph as i
...
+connecting the two vertices to the common neighbor. The number of distinct neighbors for pairs of vertices is computed
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: distinct _common_ neighbors?

File: docs/apis/batch/libs/gelly.md [hunk 3]
```diff
@@ -2051,6 +2052,26 @@ The algorithm takes a directed, vertex (and possibly edge) attributed graph as i
...
+by storing the sum of degrees of the vertex pair and subtracting the count of common neighbors, which are double-counted
+in the sum of degrees.
+
+The algorithm first annotates each edge with the endpoint degree. Grouping on the midpoint vertex, each pair of
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: Source, target or both endpoints? What is a "midpoint" vertex? Is this standard terminology? It might help to give a small example here.

File: docs/apis/batch/libs/gelly.md [hunk 4]
```diff
@@ -2051,6 +2052,26 @@ The algorithm takes a directed, vertex (and possibly edge) attributed graph as i
...
+neighbors is emitted with the endpoint degree sum. Grouping on two-paths, the common neighbors are counted.
+
+#### Usage
+The algorithm takes a simple, undirected graph as input and outputs a `DataSet` of tuples containing two vertex IDs,
+the number of common neighbors, and the number of distinct neighbors. The graph ID type must be `Comparable` and
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: Why doesn't the algorithm return the jaccard index?

File: docs/apis/batch/libs/gelly.md (hunk 1 above)
Comment (greghogan) [inline docs/apis/batch/libs/gelly.md]: I've also seen triad to mean any three vertices and triplet to mean any three connected vertices. I'll rework this section.

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+import org.apache.flink.types.IntValue;
+import org.apache.flink.types.LongValue;
+import org.apache.flink.types.NullValue;
+
+import java.text.NumberFormat;
+
+/**
+ * Driver for the library implementation of Jaccard Index.
+ *
+ * This example generates an undirected RMat graph with the given scale and
```
Comment (vasia) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java]: The rest of the examples in Gelly (and all other Flink components) are designed so that they can be run both with user-specified inputs and without (default data). Could you please modify this example to work in a similar way?

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		if (scale > 32) {
+			js = graph
+				.run(new org.apache.flink.graph.library.similarity.JaccardIndex<LongValue, NullValue, NullValue>());
+		} else {
+			js = graph
+				.run(new TranslateGraphIds<LongValue, IntValue, NullValue, NullValue>(new LongValueToIntValue()))
+				.run(new org.apache.flink.graph.library.similarity.JaccardIndex<IntValue, NullValue, NullValue>());
+		}
+
+		switch (parameters.get("output", "")) {
```
Comment (vasia) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java]: Can you add a description for these output parameter options in the class Javadoc?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/TranslateEdgeDegreeToIntValue.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.flink.types.IntValue;
+import org.apache.flink.types.LongValue;
+
+/**
+ * Translate the edge degree returned by the degree annotation functions from
+ * {@link LongValue} to {@link IntValue}.
+ *
+ * @param <T> edge value type
+ */
+public class TranslateEdgeDegreeToIntValue<T>
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/TranslateEdgeDegreeToIntValue.java]: This is a general translator from `LongValue` to `IntValue`, not just for edge degrees, right? Maybe we could rename it to demonstrate that?

File: docs/apis/batch/libs/gelly.md (hunk 4 above)
Comment (greghogan) [inline docs/apis/batch/libs/gelly.md]: It does, from `Result.getJaccardIndexScore()`.

Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: Ah great! Can you add this here?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+
+	/**
+	 * Override the default group size for the quadratic expansion of neighbor
+	 * pairs. Small groups generate more data whereas large groups distribute
+	 * computation less evenly among tasks.
+	 *
+	 * @param groupSize the group size for the quadratic expansion of neighbor pairs
+	 * @return this
+	 */
+	public JaccardIndex<K, VV, EV> setGroupSize(int groupSize) {
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: Could you maybe extend the description of what groupSize does? And it would be nice to add documentation for this method in the library docs.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+
+	/**
+	 * Filter out Jaccard Index scores less than the given minimum fraction.
+	 *
+	 * @param numerator numerator of the minimum score
+	 * @param denominator denominator of the minimum score
+	 * @return this
+	 * @see #setMaximumScore(int, int)
+	 */
+	public JaccardIndex<K, VV, EV> setMinimumScore(int numerator, int denominator) {
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: This should be documented in the library usage docs too.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+
+	/**
+	 * Filter out Jaccard Index scores greater than or equal to the given maximum fraction.
+	 *
+	 * @param numerator numerator of the maximum score
+	 * @param denominator denominator of the maximum score
+	 * @return this
+	 * @see #setMinimumScore(int, int)
+	 */
+	public JaccardIndex<K, VV, EV> setMaximumScore(int numerator, int denominator) {
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: This one also and all other configuration methods.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+	/*
+	 * Implementation notes:
+	 *
+	 * The requirement that "K extends CopyableValue<K>" can be removed when
+	 *   Flink has a self-join which performs the skew distribution handled by
+	 *   GenerateGroupSpans / GenerateGroups / GenerateGroupPairs.
+	 */
+
+	@Override
+	public DataSet<Result<K>> run(Graph<K, VV, EV> input)
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: Can you please add Javadoc for the `Result` type?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+	 * is (n choose 2) which is quadratic in the vertex degree.
+	 * <br/>
+	 * The third operation, {@link GenerateGroupPairs}, processes groups of size
+	 * {@link #groupSize} and emits {@code O(groupSize * deg(vertex))} pairs.
+	 * <br/>
+	 * This input to the third operation is still quadratic in the vertex degree.
+	 * Two prior operations, {@link GenerateGroupSpans} and {@link GenerateGroups},
+	 * each emit datasets linear in the vertex degree, with a forced rebalance
+	 * in between. {@link GenerateGroupSpans} first annotates each edge with the
+	 * number of groups and {@link GenerateGroups} emits each edge into each group.
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: 👍 for the detailed comment!

File: flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/library/similarity/JaccardIndexTest.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+			.generate();
+
+		DataSet<Result<LongValue>> ji = graph
+			.run(new JaccardIndex<LongValue, NullValue, NullValue>()
+				.setGroupSize(4));
+
+		ChecksumHashCode checksum = DataSetUtils.checksumHashCode(ji);
+
+		assertEquals(13954, checksum.getCount());
+		assertEquals(0x0000179f83a2a873L, checksum.getChecksum());
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/library/similarity/JaccardIndexTest.java]: How did you compute these values?

Comment (greghogan) [inline flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/library/similarity/JaccardIndexTest.java]: By running the algorithm.

Comment (vasia) [inline flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/library/similarity/JaccardIndexTest.java]: Then, what does you test here?

Comment (greghogan) [inline flink-libraries/flink-gelly/src/test/java/org/apache/flink/graph/library/similarity/JaccardIndexTest.java]: These are regression tests and cover aspects of the algorithms not tested by hand-crafted inputs.

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java (hunk 6 above)
Comment (greghogan) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java]: Is the usage statement sufficient?

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java (hunk 8 above)
Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: I don't know that 64 is the best value but I think it is close. I clarified the javadoc to suggest that users not change this value.

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/TranslateEdgeDegreeToIntValue.java (hunk 7 above)
Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/TranslateEdgeDegreeToIntValue.java]: I thought on this for a while, and decided that the best option is to subsume the translation in the next operation, and discovered that I had already made that change. So the class just needs to be removed.

**Comment by greghogan** [comment]
@vasia all updates should be in place.

File: docs/apis/batch/libs/gelly.md [hunk 14]
```diff
@@ -2055,22 +2055,22 @@ vertex and edge in the output graph stores the common group value and the number
 ### Jaccard Index
 
 #### Overview
-The Jaccard Index measures the similarity between vertex neighborhoods. Scores range from 0.0 (no common neighbors) to
-1.0 (all neighbors are common).
+The Jaccard Index measures the similarity between vertex neighborhoods and is computed as the number of shared numbers
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: shared numbers -> shared *neighbors?

File: docs/apis/batch/libs/gelly.md [hunk 15]
```diff
@@ -2250,14 +2250,33 @@ graph.run(new TranslateVertexValues(new LongValueAddOffset(vertexCount)));
...
 {% highlight java %}
 graph.run(new TranslateEdgeValues(new Nullify()));
 {% endhighlight %}
       </td>
     </tr>
+
+    <tr>
+      <td>library.similarity.<br/><strong>JaccardIndex</strong></td>
+      <td>
+        <p>Measures the similarity between vertex neighborhoods. The Jaccard Index score  is computed as the number of shared numbers divided by the number of distinct neighbors. Scores range from 0.0 (no shared neighbors) to 1.0 (all neighbors are shared).</p>
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: Why did you add this here and not in the "Usage" section of the library method?
I find it a bit confusing... You describe graph algorithms as building blocks for other algorithms. Does Jaccard index fall in this category?

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java [hunk 16]
```diff
@@ -40,9 +43,9 @@
 /**
  * Driver for the library implementation of Jaccard Index.
  *
- * This example generates an undirected RMat graph with the given scale and
- * edge factor then calculates all non-zero Jaccard Index similarity scores
- * between vertices.
+ * This example reads a simple, undirected graph from a CSV file or generates
```
Comment (vasia) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java]: remove one "generates"

File: flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java [hunk 17]
```diff
@@ -54,73 +57,112 @@
 
 	public static final boolean DEFAULT_CLIP_AND_FLIP = true;
 
+	private static void printUsage() {
+		System.out.println(WordUtils.wrap("The Jaccard Index measures the similarity between vertex" +
+			" neighborhoods and is computed as the number of shared numbers divided by the number of" +
```
Comment (vasia) [inline flink-libraries/flink-gelly-examples/src/main/java/org/apache/flink/graph/examples/JaccardIndex.java]: numbers -> neighbors

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java [hunk 18]
```diff
@@ -43,11 +43,13 @@
 import java.util.List;
 
 /**
- * The Jaccard Index measures the similarity between vertex neighborhoods.
- * Scores range from 0.0 (no common neighbors) to 1.0 (all neighbors are common).
+ * The Jaccard Index measures the similarity between vertex neighborhoods and
+ * is computed as the number of shared numbers divided by the number of
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/similarity/JaccardIndex.java]: numbers -> neighbors

File: docs/apis/batch/libs/gelly.md [hunk 19]
```diff
@@ -2055,22 +2055,22 @@ vertex and edge in the output graph stores the common group value and the number
...
-neighbors is emitted with the endpoint degree sum. Grouping on two-paths, the common neighbors are counted.
+The algorithm first annotates each edge with the target vertex's degree. Grouping on the source vertex, each pair of
+neighbors is emitted with the degree sum. Grouping on two-paths, the shared neighbors are counted.
 
 #### Usage
 The algorithm takes a simple, undirected graph as input and outputs a `DataSet` of tuples containing two vertex IDs,
-the number of common neighbors, and the number of distinct neighbors. The graph ID type must be `Comparable` and
-`Copyable`.
+the number of shared neighbors, and the number of distinct neighbors. The result class provides a method to compute the
+Jaccard Index score. The graph ID type must be `Comparable` and `Copyable`.
```
Comment (vasia) [inline docs/apis/batch/libs/gelly.md]: Here we should also document what is the output of the algorithm, i.e. the `Result` type and how to get the jaccard similarity out of it.

**Comment by vasia** [comment]
Thanks for the update @greghogan. I left a few comments for improvements of the docs. After these are addressed it should be good to merge.

**Comment by greghogan** [comment]
Thanks for the reviews @vasia! Merging ...
